From 176620ad3508af210128ee6e452b73935ca9a57c Mon Sep 17 00:00:00 2001
From: bootchk <bootchk@users.noreply.github.com>
Date: Sun, 22 Nov 2020 09:45:05 -0500
Subject: [PATCH 2/4] #5918 PDB file-jpeg-save critical non-interactive

---
 plug-ins/file-jpeg/jpeg.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/plug-ins/file-jpeg/jpeg.c b/plug-ins/file-jpeg/jpeg.c
index bcae891e88..f6eb6ee6d4 100644
--- a/plug-ins/file-jpeg/jpeg.c
+++ b/plug-ins/file-jpeg/jpeg.c
@@ -487,10 +487,10 @@ jpeg_save (GimpProcedure        *procedure,
       jsvals.optimize    = GIMP_VALUES_GET_BOOLEAN (args, 2);
       jsvals.progressive = GIMP_VALUES_GET_BOOLEAN (args, 3);
       image_comment      = GIMP_VALUES_DUP_STRING  (args, 4);
-      jsvals.subsmp      = GIMP_VALUES_GET_DOUBLE  (args, 5);
-      jsvals.baseline    = GIMP_VALUES_GET_DOUBLE  (args, 6);
-      jsvals.restart     = GIMP_VALUES_GET_DOUBLE  (args, 7);
-      jsvals.dct         = GIMP_VALUES_GET_DOUBLE  (args, 8);
+      jsvals.subsmp      = GIMP_VALUES_GET_INT     (args, 5);
+      jsvals.baseline    = GIMP_VALUES_GET_BOOLEAN (args, 6);
+      jsvals.restart     = GIMP_VALUES_GET_INT     (args, 7);
+      jsvals.dct         = GIMP_VALUES_GET_INT     (args, 8);
       jsvals.preview     = FALSE;
       break;
 
-- 
2.17.1

