diff --git a/plug-ins/script-fu/scripts/difference-clouds.scm b/plug-ins/script-fu/scripts/difference-clouds.scm
index 15b8fc81b9..da02760dd3 100644
--- a/plug-ins/script-fu/scripts/difference-clouds.scm
+++ b/plug-ins/script-fu/scripts/difference-clouds.scm
@@ -51,7 +51,7 @@
     (set! offset-y (+ draw-offset-y sel-offset-y))
 
     ; Offset the clouds layer
-    (if (gimp-item-is-layer drawable)
+    (if (gimp-item-id-is-layer drawable)
       (gimp-item-transform-translate diff-clouds offset-x offset-y))
 
     ; Show the solid noise dialog
