diff --git a/plug-ins/script-fu/scripts/paste-as-pattern.scm b/plug-ins/script-fu/scripts/paste-as-pattern.scm
index 01381a2c91..d2b5cb8998 100644
--- a/plug-ins/script-fu/scripts/paste-as-pattern.scm
+++ b/plug-ins/script-fu/scripts/paste-as-pattern.scm
@@ -23,7 +23,7 @@
          (pattern-draw 0)
          (path 0))
 
-    (if (= TRUE (car (gimp-image-is-valid pattern-image)))
+    (if (= TRUE (car (gimp-image-id-is-valid pattern-image)))
       (begin
         (set! pattern-draw (car (gimp-image-get-active-drawable pattern-image)))
         (set! path (string-append gimp-directory
@@ -31,13 +31,13 @@
                              filename
                              (number->string pattern-image)
                              ".pat"))
-       
+
         (file-pat-save RUN-NONINTERACTIVE
-                       pattern-image pattern-draw path path
+                       pattern-image 1 pattern-draw path
                        name)
-       
+
         (gimp-image-delete pattern-image)
-       
+
         (gimp-patterns-refresh)
         (gimp-context-set-pattern name)
       )
